ark/phenotyping/som_utils.py

```diff
@@ -73,53 +90,140 @@ def create_pixel_matrix(img_xr, seg_labels, fovs=None, channels=None, blur_facto
-        if flowsom_data is None:
-            flowsom_data = pixel_mat
-        else:
-            flowsom_data = pd.concat([flowsom_data, pixel_mat])
+        # subset the pixel matrix for training
+        pixel_mat_subset = pixel_mat.sample(frac=subset_proportion, random_state=seed)
 
-    # normalize each marker column by the 99.9 percentile value
-    flowsom_data.loc[:, channels] = flowsom_data.loc[:, channels].div(
-        flowsom_data.loc[:, channels].quantile(q=0.999, axis=0), axis=1)
+        # write complete dataset to feather, needed for cluster assignment
+        feather.write_dataframe(pixel_mat,
+                                os.path.join(base_dir,
+                                             pre_dir,
+                                             fov + ".feather"),
+                                compression='uncompressed')
 
-    return flowsom_data
+        # write subseted dataset to hdf5, needed for training
+        feather.write_dataframe(pixel_mat_subset,
+                                os.path.join(base_dir,
+                                             sub_dir,
+                                             fov + ".feather"),
+                                compression='uncompressed')
 
 
-def cluster_pixels(chan_list, base_dir,
-                   pixel_pre_name='pixel_mat_preprocessed.csv',
-                   pixel_cluster_name='pixel_mat_clustered.csv'):
-    """Run the FlowSOM training on the pixel data.
+def train_som(fovs, channels, base_dir,
+              sub_dir='pixel_mat_subsetted', weights_name='weights.feather', num_passes=1):
+    """Run the SOM training on the subsetted pixel data.
 
-    Saves results to pixel_mat_clustered.csv in base_dir.
-    Usage: Rscript som_runner.R {path_to_pixel_matrix} {chan_list_comma_separated} {save_path}
+    Saves weights to base_dir/weights_name.
 
     Args:
-        chan_list (list):
+        fovs (list):
+            The list of fovs to subset on
+        channels (list):
             The list of markers to subset on
         base_dir (str):
-            The path to the directory to save the clustered pixel matrix in
-        pixel_pre_name (str):
-            The name of the preprocessed file name, default to pixel_mat_preprocessed.csv
-        pixel_cluster_name (str):
-            The name of the file to write the clustered csv to, default to pixel_mat_clustered.csv
+            The path to the data directory
+        sub_dir (str):
+            The name of the subsetted data directory
+        weights_name (str):
+            The name of the weights file
+        num_passes (int):
+            The number of training passes to make through the dataset
     """
 
-    # set the paths to the preprocessed matrix and clustered matrix
-    preprocessed_path = os.path.join(base_dir, pixel_pre_name)
-    clustered_path = os.path.join(base_dir, pixel_cluster_name)
+    subsetted_path = os.path.join(base_dir, sub_dir)
```

Comment:
We should verify that all of the provided FOVs exist in the specified folder, and that all of the provided channels exist in one of the example processed files